data/posts/2017-06-23-expanding-on-phalcon-3-2-0-new-features.md

```diff
@@ -0,0 +1,328 @@
+## Expanding on Phalcon 3.2.0 new features
+
+With [Phalcon 3.2.0](https://blog.phalconphp.com/post/phalcon-3-2-0-released-and-new-docs) there were many new features and bugs fixed. Today we will write about most important things you need to know and show some code examples of new features.
+
+- ##### Added Factory Adapter loaders [#11001](https://github.com/phalcon/cphalcon/issues/11001)
+  With this feature you can load your services in more simplified way using for example ini file:
+  
+  ```ini
+  [database]
+  host = TEST_DB_MYSQL_HOST
+  username = TEST_DB_MYSQL_USER
+  password = TEST_DB_MYSQL_PASSWD
+  dbname = TEST_DB_MYSQL_NAME
+  port = TEST_DB_MYSQL_PORT
+  charset = TEST_DB_MYSQL_CHARSET
+  adapter = mysql
+  ```
+  
+  ```php
+  <?php
+  
+  use Phalcon\Config\Adapter\Ini;
+  use Phalcon\Di;
+  use Phalcon\Db\Adapter\Pdo\Factory;
+
+  $di = new Di();
+  $config = new Ini('config.ini');
+  
+  $di->set('config', $config);
+  
+  $di->set('db', function() {
+      return Factory::load($this->config->database);
+  });
+  ```
+  
+  This properly create your database instance and you can change database adapter to other in ini file anytime without changing your code.
+  
+- ##### Added ability to sanitize URL to `Phalcon\Filter`
+  There were added new sanitize filter - `url` which allows you to clear your urls. For example:
+  ```php
+  <?php
+
+  use Phalcon\Filter;
+
+  $filter = new Filter();
+  $wrongUrl = 'http://juhara��.co�m';
+  $correctUrl = $filter->sanitize($wrongUrl, 'url');
+  echo $correctUrl; // displays 'http://juhara.com'
+  ```
+
+- ##### Added `Phalcon\Mvc\Model::hasUpdated` and `Phalcon\Mvc\Model:getUpdatedFields`, way to check if fields were updated after create/save/update, Added option to disable snapshot update on create/save using `Phalcon\Mvc\Model::setup(['updateSnapshotOnSave' => false])` or `phalcon.orm.update_snapshot_on_save = 0` in `php.ini`
+  In Phalcon 3.1.0 we changed behavior of snapshots to be updated on model creation/update. This could potentially cause a problems to your application if you were executing `Model::getChangedFields` in `afterUpdate`, `afterSave` or `afterCreate`. This change was done to fix other things(for example dynamic update). Right now you have two options - change your methods from `hasChanged`, `getChangedFields`, `getSnapshot` to `hasUpdated`, `getUpdatedFiels`, `getOldSnapshot` or add `phalcon.orm.update_snapshot_on_save = 0` to your `php.ini` to disable snapshot updating on save.
```

Comment:
In Phalcon 3.1.0 we changed the behavior of snapshots; they are now updated on model creation/update. This could potentially cause problems to your application if you execute `Model::getChangedFields` in `afterUpdate`, `afterSave` or `afterCreate`. This change was done to fix other things also (for example dynamic update). Right now you have two options - change your methods from `hasChanged`, `getChangedFields`, `getSnapshot` to `hasUpdated`, `getUpdatedFiels`, `getOldSnapshot` or add `phalcon.orm.update_snapshot_on_save = 0` to your `php.ini` to disable snapshot updating on save.